src/multipart/multipart.rs

```diff
@@ -0,0 +1,471 @@
+extern crate url;
+extern crate uuid;
+
+use std;
+use std::borrow::Cow;
+use std::io::Read;
+use hyper::mime::Mime;
+
+/// A multipart/form-data request.
+#[derive(Debug)]
+pub struct MultipartRequest {
+    boundary: String,
+    fields: Vec<MultipartField>,
+}
+
+impl MultipartRequest {
+    /// Creates a new MultipartRequest without any content.
+    pub fn new() -> MultipartRequest {
+        MultipartRequest {
+            boundary: format!("{}", uuid::Uuid::new_v4().simple()),
+            fields: Vec::new(),
+        }
+    }
+
+    /// Adds a field, builder style.
+    pub fn field(mut self, field: MultipartField) -> MultipartRequest {
+        self.fields.push(field);
+        self
+    }
+
+    /// Adds multiple fields.
+    pub fn fields(&mut self, mut fields: Vec<MultipartField>) {
+        self.fields.append(&mut fields);
+    }
+
+    fn reader(self) -> MultipartReader {
+        MultipartReader::new(self)
+    }
+
+    /// Gets the automatically chosen boundary.
+    pub fn boundary(&self) -> &str {
+        &self.boundary
+    }
+
+    fn compute_length(&self) -> Option<u64> {
+        let mut length = 0u64;
+        for field in self.fields.iter() {
+            match field.value_length {
+                Some(value_length) => {
+                    // We are constructing the header just to get its length. This is wasteful
+                    // but it seemed liked the only way to get the correct length for sure.
+                    // TODO: Should we instead cache the computed header for when it is used again
+                    // when sending the request? This would use more memory as all headers would be
+                    // held in memory as opposed to only the current one but it would only compute
+                    // each header once.
```

Comment:
Oh, I'm not worried either way. I suppose for this to user a whole lot of memory, there would need to be **a lot** of params, which is unlikely, and the memory needed for these strings is probably tiny in comparison. So, I suppose caching them would be an improvement.